File: Harvest/Services/CommunityService.swift
import Foundation
import Supabase
import Realtime

struct CommunityService {
    private var client: SupabaseClient { SupabaseManager.shared.client }

    /// Rooms the user is allowed to join (via the access-rules RPC).
    func availableCommunities(userId: String) async throws -> [Community] {
        try await client
            .rpc("available_communities", params: ["p_user": userId])
            .execute()
            .value
    }

    /// Community ids the user has actively joined.
    func joinedCommunityIds(userId: String) async throws -> Set<String> {
        struct Row: Decodable { let community_id: String }
        let rows: [Row] = try await client
            .from("community_members")
            .select("community_id")
            .eq("user_id", value: userId)
            .eq("status", value: "active")
            .execute()
            .value
        return Set(rows.map(\.community_id))
    }

    func join(communityId: String, userId: String) async throws {
        try await client
            .from("community_members")
            .upsert([
                "community_id": communityId,
                "user_id": userId,
                "status": "active"
            ])
            .execute()
    }

    func leave(communityId: String, userId: String) async throws {
        try await client
            .from("community_members")
            .update(["status": "left"])
            .eq("community_id", value: communityId)
            .eq("user_id", value: userId)
            .execute()
    }

    /// Latest page of messages, newest-first. Pass the oldest loaded
    /// created_at as `before` to fetch the next older page.
    func messagesPage(communityId: String, before: String? = nil, limit: Int = 50) async throws -> [CommunityMessage] {
        var query = client
            .from("community_messages")
            .select()
            .eq("community_id", value: communityId)
            .eq("is_removed", value: false)
        if let before {
            query = query.lt("created_at", value: before)
        }
        return try await query
            .order("created_at", ascending: false)
            .limit(limit)
            .execute()
            .value
    }

    /// Fetch specific messages by id — used for quoted-reply previews whose
    /// originals fell outside the loaded pages. Includes removed rows so the
    /// UI can render "Message removed".
    func messagesByIds(_ ids: [String]) async throws -> [CommunityMessage] {
        guard !ids.isEmpty else { return [] }
        return try await client
            .from("community_messages")
            .select()
            .in("id", values: ids)
            .execute()
            .value
    }

    private struct NewCommunityMessage: Encodable {
        let community_id: String
        let sender_id: String
        let content: String
        let reply_to_id: String?
        let mentions: [String]
    }

    /// Throws ContactInfoBlocked when server-side detection rejects the message.
    /// Returns the inserted row so the sender sees the message immediately,
    /// without waiting for the realtime echo.
    @discardableResult
    func post(
        communityId: String,
        senderId: String,
        content: String,
        replyToId: String? = nil,
        mentions: [String] = []
    ) async throws -> CommunityMessage? {
        let inserted: [CommunityMessage] = try await client
            .from("community_messages")
            .insert(NewCommunityMessage(
                community_id: communityId,
                sender_id: senderId,
                content: content,
                reply_to_id: replyToId,
                mentions: mentions
            ))
            .select()
            .execute()
            .value
        return inserted.first
    }

    /// Name + avatar for the given user ids (for chat bubbles).
    func senderProfiles(ids: [String]) async throws -> [CommunitySender] {
        guard !ids.isEmpty else { return [] }
        return try await client
            .from("users")
            .select("id, nickname, photos")
            .in("id", values: ids)
            .execute()
            .value
    }

    /// All reactions for the given messages (bulk, one query per page load).
    func reactions(messageIds: [String]) async throws -> [CommunityReaction] {
        guard !messageIds.isEmpty else { return [] }
        return try await client
            .from("community_message_reactions")
            .select()
            .in("message_id", values: messageIds)
            .execute()
            .value
    }

    func addReaction(messageId: String, userId: String, emoji: String) async throws {
        // community_id is filled by a DB trigger — never send it.
        try await client
            .from("community_message_reactions")
            .upsert([
                "message_id": messageId,
                "user_id": userId,
                "emoji": emoji
            ])
            .execute()
    }

    func removeReaction(messageId: String, userId: String, emoji: String) async throws {
        try await client
            .from("community_message_reactions")
            .delete()
            .eq("message_id", value: messageId)
            .eq("user_id", value: userId)
            .eq("emoji", value: emoji)
            .execute()
    }

    /// Active members of a room (for @mention autocomplete).
    func members(communityId: String) async throws -> [CommunitySender] {
        struct Row: Decodable { let users: CommunitySender }
        let rows: [Row] = try await client
            .from("community_members")
            .select("users(id, nickname, photos)")
            .eq("community_id", value: communityId)
            .eq("status", value: "active")
            .execute()
            .value
        return rows.map(\.users)
    }

    func prompts(communityId: String) async throws -> [CommunityPrompt] {
        // Room-specific OR global (community_id is null).
        try await client
            .from("community_prompts")
            .select("id, text")
            .or("community_id.eq.\(communityId),community_id.is.null")
            .eq("is_active", value: true)
            .order("display_order", ascending: true)
            .execute()
            .value
    }

    func subscribe(communityId: String, onMessage: @escaping @Sendable (CommunityMessage) -> Void) -> RealtimeChannelV2 {
        let channel = client.realtimeV2.channel("community:\(communityId)")
        let changes = channel.postgresChange(
            InsertAction.self,
            table: "community_messages",
            filter: .eq("community_id", value: communityId)
        )
        Task {
            for await change in changes {
                if let msg = try? change.decodeRecord(as: CommunityMessage.self, decoder: JSONDecoder()) {
                    onMessage(msg)
                }
            }
        }
        Task { try? await channel.subscribeWithError() }
        return channel
    }

    /// Live reaction add/remove events for one room. DELETE events carry the
    /// full old row because the table has replica identity full.
    func subscribeReactions(
        communityId: String,
        onInsert: @escaping @Sendable (CommunityReaction) -> Void,
        onDelete: @escaping @Sendable (CommunityReaction) -> Void
    ) -> RealtimeChannelV2 {
        let channel = client.realtimeV2.channel("community-reactions:\(communityId)")
        let inserts = channel.postgresChange(
            InsertAction.self,
            table: "community_message_reactions",
            filter: .eq("community_id", value: communityId)
        )
        let deletes = channel.postgresChange(
            DeleteAction.self,
            table: "community_message_reactions",
            filter: .eq("community_id", value: communityId)
        )
        Task {
            for await change in inserts {
                if let r = try? change.decodeRecord(as: CommunityReaction.self, decoder: JSONDecoder()) {
                    onInsert(r)
                }
            }
        }
        Task {
            for await change in deletes {
                if let r = try? change.decodeOldRecord(as: CommunityReaction.self, decoder: JSONDecoder()) {
                    onDelete(r)
                }
            }
        }
        Task { try? await channel.subscribeWithError() }
        return channel
    }

    func unsubscribe(_ channel: RealtimeChannelV2) {
        Task { await channel.unsubscribe() }
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/data/service/CommunityService.kt
package com.harvestglass.harvest.data.service

import com.harvestglass.harvest.data.model.Community
import com.harvestglass.harvest.data.model.CommunityMessage
import com.harvestglass.harvest.data.model.CommunityPrompt
import com.harvestglass.harvest.data.model.CommunityReaction
import com.harvestglass.harvest.data.model.CommunitySender
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.realtime.PostgresAction
import io.github.jan.supabase.realtime.channel
import io.github.jan.supabase.realtime.decodeOldRecord
import io.github.jan.supabase.realtime.decodeRecord
import io.github.jan.supabase.realtime.postgresChangeFlow
import io.github.jan.supabase.realtime.realtime
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.flow.merge
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray

/** A live reaction event for one room. */
sealed interface ReactionEvent {
    data class Added(val reaction: CommunityReaction) : ReactionEvent
    data class Removed(val reaction: CommunityReaction) : ReactionEvent
}

/** Mirrors Harvest/Services/CommunityService.swift, method for method. */
class CommunityService(private val client: SupabaseClient) {

    // MARK: - Rooms

    /** Rooms the user is allowed to join (via the access-rules RPC). */
    suspend fun availableCommunities(userId: String): List<Community> =
        client.postgrest
            .rpc("available_communities", buildJsonObject { put("p_user", userId) })
            .decodeList<Community>()

    /** Community ids the user has actively joined. */
    suspend fun joinedCommunityIds(userId: String): Set<String> {
        @Serializable
        data class Row(val community_id: String)

        return client.postgrest.from("community_members")
            .select(Columns.list("community_id")) {
                filter {
                    eq("user_id", userId)
                    eq("status", "active")
                }
            }
            .decodeList<Row>()
            .map { it.community_id }
            .toSet()
    }

    suspend fun join(communityId: String, userId: String) {
        client.postgrest.from("community_members").upsert(
            buildJsonObject {
                put("community_id", communityId)
                put("user_id", userId)
                put("status", "active")
            }
        )
    }

    suspend fun leave(communityId: String, userId: String) {
        client.postgrest.from("community_members").update(
            buildJsonObject { put("status", "left") }
        ) {
            filter {
                eq("community_id", communityId)
                eq("user_id", userId)
            }
        }
    }

    // MARK: - Messages

    /**
     * Latest page of messages, newest-first. Pass the oldest loaded
     * created_at as [before] to fetch the next older page.
     */
    suspend fun messagesPage(
        communityId: String,
        before: String? = null,
        limit: Int = 50
    ): List<CommunityMessage> =
        client.postgrest.from("community_messages")
            .select {
                filter {
                    eq("community_id", communityId)
                    eq("is_removed", false)
                    if (before != null) lt("created_at", before)
                }
                order("created_at", Order.DESCENDING)
                limit(limit.toLong())
            }
            .decodeList()

    /**
     * Fetch specific messages by id — used for quoted-reply previews whose
     * originals fell outside the loaded pages. Includes removed rows so the
     * UI can render "Message removed".
     */
    suspend fun messagesByIds(ids: List<String>): List<CommunityMessage> {
        if (ids.isEmpty()) return emptyList()
        return client.postgrest.from("community_messages")
            .select { filter { isIn("id", ids) } }
            .decodeList()
    }

    /**
     * Returns the inserted row so the sender sees the message immediately,
     * without waiting for the realtime echo. Throws when server-side contact
     * detection rejects the message.
     */
    suspend fun post(
        communityId: String,
        senderId: String,
        content: String,
        replyToId: String? = null,
        mentions: List<String> = emptyList()
    ): CommunityMessage? =
        client.postgrest.from("community_messages")
            .insert(
                buildJsonObject {
                    put("community_id", communityId)
                    put("sender_id", senderId)
                    put("content", content)
                    put("reply_to_id", replyToId)
                    putJsonArray("mentions") { mentions.forEach { add(it) } }
                }
            ) { select() }
            .decodeList<CommunityMessage>()
            .firstOrNull()

    // MARK: - People

    /** Name + avatar for the given user ids (for chat bubbles). */
    suspend fun senderProfiles(ids: List<String>): List<CommunitySender> {
        if (ids.isEmpty()) return emptyList()
        return client.postgrest.from("users")
            .select(Columns.raw("id, nickname, photos")) { filter { isIn("id", ids) } }
            .decodeList()
    }

    /** Active members of a room (for @mention autocomplete). */
    suspend fun members(communityId: String): List<CommunitySender> {
        @Serializable
        data class Row(val users: CommunitySender)

        return client.postgrest.from("community_members")
            .select(Columns.raw("users(id, nickname, photos)")) {
                filter {
                    eq("community_id", communityId)
                    eq("status", "active")
                }
            }
            .decodeList<Row>()
            .map { it.users }
    }

    // MARK: - Reactions

    /** All reactions for the given messages (bulk, one query per page load). */
    suspend fun reactions(messageIds: List<String>): List<CommunityReaction> {
        if (messageIds.isEmpty()) return emptyList()
        return client.postgrest.from("community_message_reactions")
            .select { filter { isIn("message_id", messageIds) } }
            .decodeList()
    }

    suspend fun addReaction(messageId: String, userId: String, emoji: String) {
        // community_id is filled by a DB trigger — never send it.
        client.postgrest.from("community_message_reactions").upsert(
            buildJsonObject {
                put("message_id", messageId)
                put("user_id", userId)
                put("emoji", emoji)
            }
        )
    }

    suspend fun removeReaction(messageId: String, userId: String, emoji: String) {
        client.postgrest.from("community_message_reactions").delete {
            filter {
                eq("message_id", messageId)
                eq("user_id", userId)
                eq("emoji", emoji)
            }
        }
    }

    // MARK: - Prompts

    suspend fun prompts(communityId: String): List<CommunityPrompt> =
        // Room-specific OR global (community_id is null).
        client.postgrest.from("community_prompts")
            .select(Columns.raw("id, text")) {
                filter {
                    or {
                        eq("community_id", communityId)
                        exact("community_id", null)
                    }
                    eq("is_active", true)
                }
                order("display_order", Order.ASCENDING)
            }
            .decodeList()

    // MARK: - Realtime

    /** Live inserts for one room, server-filtered by community_id. */
    fun subscribeMessages(communityId: String): Flow<CommunityMessage> = flow {
        val channel = client.realtime.channel("community:$communityId")
        val changes = channel.postgresChangeFlow<PostgresAction.Insert>(schema = "public") {
            table = "community_messages"
            filter("community_id", io.github.jan.supabase.postgrest.query.filter.FilterOperator.EQ, communityId)
        }
        channel.subscribe()
        emitAll(changes.mapNotNull { runCatching { it.decodeRecord<CommunityMessage>() }.getOrNull() })
    }

    /**
     * Live reaction add/remove events for one room. INSERT events are
     * server-filtered by community_id; DELETE events are not filterable in
     * postgres_changes and (with RLS) carry only the primary-key columns —
     * that's sufficient, since removal matches on (messageId, userId, emoji)
     * and events for unloaded messages no-op.
     */
    fun subscribeReactions(communityId: String): Flow<ReactionEvent> = flow {
        val channel = client.realtime.channel("community-reactions:$communityId")
        val inserts = channel.postgresChangeFlow<PostgresAction.Insert>(schema = "public") {
            table = "community_message_reactions"
            filter("community_id", io.github.jan.supabase.postgrest.query.filter.FilterOperator.EQ, communityId)
        }
        val deletes = channel.postgresChangeFlow<PostgresAction.Delete>(schema = "public") {
            table = "community_message_reactions"
        }
        channel.subscribe()
        emitAll(
            merge(
                inserts.mapNotNull {
                    runCatching { ReactionEvent.Added(it.decodeRecord<CommunityReaction>()) }.getOrNull()
                },
                deletes.mapNotNull {
                    runCatching { ReactionEvent.Removed(it.decodeOldRecord<CommunityReaction>()) }.getOrNull()
                }
            )
        )
    }
}
